File: c/gtk3/app-chooser-dialog.c
#include <gtk/gtk.h>

static void on_app_activate(GApplication* self, gpointer data);
static void on_app_startup(GApplication* self, gpointer data);

const static gchar* APP_ID = "io.github.Miqueas.GTK-Examples.C.Gtk3.AppChooserDialog";
const static gchar* APP_TITLE = "GtkAppChooserDialog";

gint main(gint argc, gchar** argv) {
  GtkApplication* app = gtk_application_new(APP_ID, 0);
  g_signal_connect(app, "startup", G_CALLBACK(on_app_startup), NULL);
  g_signal_connect(app, "activate", G_CALLBACK(on_app_activate), NULL);

  gint result = g_application_run(G_APPLICATION(app), argc, argv);
  g_object_unref(app);

  return result;
}

static void on_app_activate(GApplication* self, gpointer data) {
  GtkWindow* window = gtk_application_get_active_window(GTK_APPLICATION(self));

  if (window == NULL) {
    g_printerr("No active window found.\n");
    return;
  }

  gint result = gtk_dialog_run(GTK_DIALOG(window));

  switch (result) {
    case GTK_RESPONSE_OK: {
      GAppInfo* info = gtk_app_chooser_get_app_info(GTK_APP_CHOOSER(window));
      g_print("ID: %s\n", g_app_info_get_id(info));
      g_print("Executable: %s\n", g_app_info_get_executable(info));
      g_print("Description: %s\n", g_app_info_get_description(info));
      g_print("Commandline: %s\n", g_app_info_get_commandline(info));
      g_print("Display name: %s\n", g_app_info_get_display_name(info));
      break;
    }
    default: break;
  }

  gtk_widget_destroy(GTK_WIDGET(window));
}

static void on_app_startup(GApplication* self, gpointer data) {
  GtkWidget* dialog = gtk_app_chooser_dialog_new_for_content_type(NULL, 0, "image/png");

  gtk_window_set_title(GTK_WINDOW(dialog), APP_TITLE);
  gtk_window_set_application(GTK_WINDOW(dialog), GTK_APPLICATION(self));
  gtk_container_set_border_width(GTK_CONTAINER(dialog), 10);
}

File: zig/gtk3/app-chooser-dialog.zig
const std = @import("std");
const gtk = @import("gtk");
const gio = @import("gio");
const gobject = @import("gobject");

const APP_ID = "io.github.Miqueas.GTK-Examples.Zig.Gtk3.AppChooserDialog";
const APP_TITLE = "GtkAppChooserDialog";

pub fn main(init: std.process.Init) void {
    var app = gtk.Application.new(APP_ID, .{});
    defer app.unref();

    _ = gio.Application.signals.startup.connect(app, ?*anyopaque, &onStartup, null, .{});
    _ = gio.Application.signals.activate.connect(app, ?*anyopaque, &onActivate, null, .{});

    const argv = init.minimal.args.vector;
    const status = gio.Application.run(
        app.as(gio.Application),
        @intCast(argv.len),
        @ptrCast(@constCast(argv.ptr))
    );

    std.process.exit(@intCast(status));
}

fn onActivate(app: *gtk.Application, _: ?*anyopaque) callconv(.c) void {
    if (app.getActiveWindow() == null) return;

    const window = app.getActiveWindow().?;
    const dialog = gobject.ext.cast(gtk.Dialog, window).?;
    const result = dialog.run();

    switch (@as(gtk.ResponseType, @enumFromInt(result))) {
        .ok => {
            const app_chooser = gobject.ext.cast(gtk.AppChooser, window).?;
            const info = gtk.AppChooser.getAppInfo(app_chooser).?;
            std.debug.print("ID: {s}\n", .{ info.getId().? });
            std.debug.print("Executable: {s}\n", .{ info.getExecutable() });
            std.debug.print("Description: {s}\n", .{ info.getDescription().? });
            std.debug.print("Commandline: {s}\n", .{ info.getCommandline().? });
            std.debug.print("Display name: {s}\n", .{ info.getDisplayName() });
        },
        else => {},
    }

    gtk.Widget.destroy(window.as(gtk.Widget));
}

fn onStartup(app: *gtk.Application, _: ?*anyopaque) callconv(.c) void {
    const dialog = gtk.AppChooserDialog.newForContentType(null, .{}, "image/png");

    gtk.Window.setTitle(dialog.as(gtk.Window), APP_TITLE);
    gtk.Window.setApplication(dialog.as(gtk.Window), app);
    gtk.Container.setBorderWidth(dialog.as(gtk.Container), 10);
}